### PR TITLE
Fix Tracker AD for LDA 

### LINKED ISSUE #850: Tracker AD is broken for LDA
Running the following model on `master` would produce an AD error

```julia
# Simulate data
using Distributions, Turing

Turing.setadbackend(:reverse_diff)

V = 5   # vocab
K = 3   # topic
M = 25  # doc

alpha = collect(ones(K) / K);
beta = collect(ones(V) / V);

theta = rand(Dirichlet(alpha), M)
phi = rand(Dirichlet(beta), K)
@info "true vals" phi theta sum(phi * theta, dims=1)

avg_doc_length = 50

doc_length = rand(Poisson(avg_doc_length), M)

N = sum(doc_length)

w = Vector{Int}(undef, N)
doc = Vector{Int}(undef, N)

n = 1
for m = 1:M, i=1:doc_length[m]
    global n
    z = rand(Categorical(theta[:,m]))
    w[n] = rand(Categorical(phi[:,z]))
    doc[n] = m
    n = n + 1
end

# LDA

struct CategoricalReal<: Distributions.DiscreteUnivariateDistribution
    p
end

function Distributions.logpdf(d::CategoricalReal, k::Int64)
    return log(d.p[k])
end

@model lda_collapsed(K, V, M, N, w, doc, beta, alpha, ::Type{T}=Float64) where {T<:Real} = begin
    theta = Matrix{T}(undef, K, M)
    for m = 1:M
        theta[:,m] ~ Dirichlet(alpha)
    end

    phi = Matrix{T}(undef, V, K)
    for k = 1:K
        phi[:,k] ~ Dirichlet(beta)
    end

    word_dist = phi * theta

    for n = 1:N
        w[n] ~ CategoricalReal(word_dist[:,doc[n]])
    end
end

mf = lda_collapsed(K, V, M, N, w, doc, beta, alpha)

c3 = sample(mf, HMC(2000, 0.01, 1));
```

Stack trace:

```julia

julia> c3 = sample(mf, HMC(2000, 0.01, 1));
ERROR: MethodError: no method matching Float64(::Tracker.TrackedReal{Float64})
Closest candidates are:
  Float64(::Real, ::RoundingMode) where T<:AbstractFloat at rounding.jl:194
  Float64(::T<:Number) where T<:Number at boot.jl:741
  Float64(::Int8) at float.jl:60
  ...
Stacktrace:
 [1] convert(::Type{Float64}, ::Tracker.TrackedReal{Float64}) at ./number.jl:7
 [2] setindex!(::Array{Float64,1}, ::Tracker.TrackedReal{Float64}, ::Int64) at ./array.jl:767
 [3] invlink(::Dirichlet{Float64}, ::TrackedArray{…,SubArray{Float64,1,Array{Float64,1},Tuple{UnitRange{Int64}},true}}, ::Type{Val{true}}) at /Users/hg344/.julia/packages/Bijectors/TsBpO/src/Bijectors.jl:231
 [4] invlink(::Dirichlet{Float64}, ::TrackedArray{…,SubArray{Float64,1,Array{Float64,1},Tuple{UnitRange{Int64}},true}}) at /Users/hg344/.julia/packages/Bijectors/TsBpO/src/Bijectors.jl:227
 [5] getindex(::Turing.Core.RandomVariables.VarInfo{NamedTuple{(:theta, :phi),Tuple{Turing.Core.RandomVariables.Metadata{Dict{Turing.Core.RandomVariables.VarName{:theta},Int64},Array{Dirichlet{Float64},1},Array{Turing.Core.RandomVariables.VarName{:theta},1},TrackedArray{…,Array{Float64,1}},Array{Set{Turing.Selector},1}},Turing.Core.RandomVariables.Metadata{Dict{Turing.Core.RandomVariables.VarName{:phi},Int64},Array{Dirichlet{Float64},1},Array{Turing.Core.RandomVariables.VarName{:phi},1},TrackedArray{…,Array{Float64,1}},Array{Set{Turing.Selector},1}}}},Tracker.TrackedReal{Float64}}, ::Turing.Core.RandomVariables.VarName{:theta}) at /Users/hg344/.julia/dev/Turing/src/core/RandomVariables.jl:983
 [6] assume at /Users/hg344/.julia/dev/Turing/src/inference/hmc.jl:612 [inlined]
 [7] macro expansion at /Users/hg344/.julia/dev/Turing/src/core/compiler.jl:101 [inlined]
 [8] macro expansion at ./REPL[21]:4 [inlined]
 [9] (::getfield(Main, Symbol("###inner_function#376#8")))(::Turing.Core.RandomVariables.VarInfo{NamedTuple{(:theta, :phi),Tuple{Turing.Core.RandomVariables.Metadata{Dict{Turing.Core.RandomVariables.VarName{:theta},Int64},Array{Dirichlet{Float64},1},Array{Turing.Core.RandomVariables.VarName{:theta},1},TrackedArray{…,Array{Float64,1}},Array{Set{Turing.Selector},1}},Turing.Core.RandomVariables.Metadata{Dict{Turing.Core.RandomVariables.VarName{:phi},Int64},Array{Dirichlet{Float64},1},Array{Turing.Core.RandomVariables.VarName{:phi},1},TrackedArray{…,Array{Float64,1}},Array{Set{Turing.Selector},1}}}},Tracker.TrackedReal{Float64}}, ::Turing.Sampler{HMC{Turing.Core.TrackerAD,(),AdvancedHMC.Adaptation.UnitEuclideanMetric}}, ::Turing.Model{Tuple{:theta,:phi},Tuple{:K,:V,:M,:N,:w,:doc,:beta,:alpha,:T},getfield(Main, Symbol("###inner_function#376#8")),NamedTuple{(:K, :V, :M, :N, :w, :doc, :beta, :alpha, :T),Tuple{Int64,Int64,Int64,Int64,Array{Int64,1},Array{Int64,1},Array{Float64,1},Array{Float64,1},Type{Float64}}},NamedTuple{(:K, :V, :M, :N, :w, :doc, :beta, :alpha),NTuple{8,Symbol}}}) at /Users/hg344/.julia/dev/Turing/src/core/compiler.jl:438
 [10] #call#3 at /Users/hg344/.julia/dev/Turing/src/Turing.jl:73 [inlined]
 [11] Model at /Users/hg344/.julia/dev/Turing/src/Turing.jl:73 [inlined]
 [12] runmodel!(::Turing.Model{Tuple{:theta,:phi},Tuple{:K,:V,:M,:N,:w,:doc,:beta,:alpha,:T},getfield(Main, Symbol("###inner_function#376#8")),NamedTuple{(:K, :V, :M, :N, :w, :doc, :beta, :alpha, :T),Tuple{Int64,Int64,Int64,Int64,Array{Int64,1},Array{Int64,1},Array{Float64,1},Array{Float64,1},Type{Float64}}},NamedTuple{(:K, :V, :M, :N, :w, :doc, :beta, :alpha),NTuple{8,Symbol}}}, ::Turing.Core.RandomVariables.VarInfo{NamedTuple{(:theta, :phi),Tuple{Turing.Core.RandomVariables.Metadata{Dict{Turing.Core.RandomVariables.VarName{:theta},Int64},Array{Dirichlet{Float64},1},Array{Turing.Core.RandomVariables.VarName{:theta},1},TrackedArray{…,Array{Float64,1}},Array{Set{Turing.Selector},1}},Turing.Core.RandomVariables.Metadata{Dict{Turing.Core.RandomVariables.VarName{:phi},Int64},Array{Dirichlet{Float64},1},Array{Turing.Core.RandomVariables.VarName{:phi},1},TrackedArray{…,Array{Float64,1}},Array{Set{Turing.Selector},1}}}},Tracker.TrackedReal{Float64}}, ::Turing.Sampler{HMC{Turing.Core.TrackerAD,(),AdvancedHMC.Adaptation.UnitEuclideanMetric}}) at /Users/hg344/.julia/dev/Turing/src/core/RandomVariables.jl:699
 [13] (::getfield(Turing.Core, Symbol("#f#32")){Turing.Core.RandomVariables.VarInfo{NamedTuple{(:theta, :phi),Tuple{Turing.Core.RandomVariables.Metadata{Dict{Turing.Core.RandomVariables.VarName{:theta},Int64},Array{Dirichlet{Float64},1},Array{Turing.Core.RandomVariables.VarName{:theta},1},Array{Float64,1},Array{Set{Turing.Selector},1}},Turing.Core.RandomVariables.Metadata{Dict{Turing.Core.RandomVariables.VarName{:phi},Int64},Array{Dirichlet{Float64},1},Array{Turing.Core.RandomVariables.VarName{:phi},1},Array{Float64,1},Array{Set{Turing.Selector},1}}}},Float64},Turing.Model{Tuple{:theta,:phi},Tuple{:K,:V,:M,:N,:w,:doc,:beta,:alpha,:T},getfield(Main, Symbol("###inner_function#376#8")),NamedTuple{(:K, :V, :M, :N, :w, :doc, :beta, :alpha, :T),Tuple{Int64,Int64,Int64,Int64,Array{Int64,1},Array{Int64,1},Array{Float64,1},Array{Float64,1},Type{Float64}}},NamedTuple{(:K, :V, :M, :N, :w, :doc, :beta, :alpha),NTuple{8,Symbol}}},Turing.Sampler{HMC{Turing.Core.TrackerAD,(),AdvancedHMC.Adaptation.UnitEuclideanMetric}}})(::TrackedArray{…,Array{Float64,1}}) at /Users/hg344/.julia/dev/Turing/src/core/ad.jl:145
 [14] #20 at /Users/hg344/.julia/packages/Tracker/RRYy6/src/back.jl:148 [inlined]
 [15] forward(::getfield(Tracker, Symbol("##20#22")){getfield(Turing.Core, Symbol("#f#32")){Turing.Core.RandomVariables.VarInfo{NamedTuple{(:theta, :phi),Tuple{Turing.Core.RandomVariables.Metadata{Dict{Turing.Core.RandomVariables.VarName{:theta},Int64},Array{Dirichlet{Float64},1},Array{Turing.Core.RandomVariables.VarName{:theta},1},Array{Float64,1},Array{Set{Turing.Selector},1}},Turing.Core.RandomVariables.Metadata{Dict{Turing.Core.RandomVariables.VarName{:phi},Int64},Array{Dirichlet{Float64},1},Array{Turing.Core.RandomVariables.VarName{:phi},1},Array{Float64,1},Array{Set{Turing.Selector},1}}}},Float64},Turing.Model{Tuple{:theta,:phi},Tuple{:K,:V,:M,:N,:w,:doc,:beta,:alpha,:T},getfield(Main, Symbol("###inner_function#376#8")),NamedTuple{(:K, :V, :M, :N, :w, :doc, :beta, :alpha, :T),Tuple{Int64,Int64,Int64,Int64,Array{Int64,1},Array{Int64,1},Array{Float64,1},Array{Float64,1},Type{Float64}}},NamedTuple{(:K, :V, :M, :N, :w, :doc, :beta, :alpha),NTuple{8,Symbol}}},Turing.Sampler{HMC{Turing.Core.TrackerAD,(),AdvancedHMC.Adaptation.UnitEuclideanMetric}}}}, ::Tracker.Params) at /Users/hg344/.julia/packages/Tracker/RRYy6/src/back.jl:135
 [16] forward(::Function, ::Array{Float64,1}) at /Users/hg344/.julia/packages/Tracker/RRYy6/src/back.jl:148
 [17] gradient_logp_reverse(::Array{Float64,1}, ::Turing.Core.RandomVariables.VarInfo{NamedTuple{(:theta, :phi),Tuple{Turing.Core.RandomVariables.Metadata{Dict{Turing.Core.RandomVariables.VarName{:theta},Int64},Array{Dirichlet{Float64},1},Array{Turing.Core.RandomVariables.VarName{:theta},1},Array{Float64,1},Array{Set{Turing.Selector},1}},Turing.Core.RandomVariables.Metadata{Dict{Turing.Core.RandomVariables.VarName{:phi},Int64},Array{Dirichlet{Float64},1},Array{Turing.Core.RandomVariables.VarName{:phi},1},Array{Float64,1},Array{Set{Turing.Selector},1}}}},Float64}, ::Turing.Model{Tuple{:theta,:phi},Tuple{:K,:V,:M,:N,:w,:doc,:beta,:alpha,:T},getfield(Main, Symbol("###inner_function#376#8")),NamedTuple{(:K, :V, :M, :N, :w, :doc, :beta, :alpha, :T),Tuple{Int64,Int64,Int64,Int64,Array{Int64,1},Array{Int64,1},Array{Float64,1},Array{Float64,1},Type{Float64}}},NamedTuple{(:K, :V, :M, :N, :w, :doc, :beta, :alpha),NTuple{8,Symbol}}}, ::Turing.Sampler{HMC{Turing.Core.TrackerAD,(),AdvancedHMC.Adaptation.UnitEuclideanMetric}}) at /Users/hg344/.julia/dev/Turing/src/core/ad.jl:151
 [18] gradient_logp at /Users/hg344/.julia/dev/Turing/src/core/ad.jl:82 [inlined]
 [19] (::getfield(Turing.Inference, Symbol("#∂logπ∂θ#30")){Turing.Core.RandomVariables.VarInfo{NamedTuple{(:theta, :phi),Tuple{Turing.Core.RandomVariables.Metadata{Dict{Turing.Core.RandomVariables.VarName{:theta},Int64},Array{Dirichlet{Float64},1},Array{Turing.Core.RandomVariables.VarName{:theta},1},Array{Float64,1},Array{Set{Turing.Selector},1}},Turing.Core.RandomVariables.Metadata{Dict{Turing.Core.RandomVariables.VarName{:phi},Int64},Array{Dirichlet{Float64},1},Array{Turing.Core.RandomVariables.VarName{:phi},1},Array{Float64,1},Array{Set{Turing.Selector},1}}}},Float64},Turing.Sampler{HMC{Turing.Core.TrackerAD,(),AdvancedHMC.Adaptation.UnitEuclideanMetric}},Turing.Model{Tuple{:theta,:phi},Tuple{:K,:V,:M,:N,:w,:doc,:beta,:alpha,:T},getfield(Main, Symbol("###inner_function#376#8")),NamedTuple{(:K, :V, :M, :N, :w, :doc, :beta, :alpha, :T),Tuple{Int64,Int64,Int64,Int64,Array{Int64,1},Array{Int64,1},Array{Float64,1},Array{Float64,1},Type{Float64}}},NamedTuple{(:K, :V, :M, :N, :w, :doc, :beta, :alpha),NTuple{8,Symbol}}}})(::Array{Float64,1}) at /Users/hg344/.julia/dev/Turing/src/inference/hmc.jl:519
 [20] ∂H∂θ at /Users/hg344/.julia/packages/AdvancedHMC/DdEnk/src/hamiltonian.jl:13 [inlined]
 [21] phasepoint(::AdvancedHMC.Hamiltonian{AdvancedHMC.Adaptation.UnitEuclideanMetric{Float64},getfield(Turing.Inference, Symbol("#logπ#31")){Turing.Core.RandomVariables.VarInfo{NamedTuple{(:theta, :phi),Tuple{Turing.Core.RandomVariables.Metadata{Dict{Turing.Core.RandomVariables.VarName{:theta},Int64},Array{Dirichlet{Float64},1},Array{Turing.Core.RandomVariables.VarName{:theta},1},Array{Float64,1},Array{Set{Turing.Selector},1}},Turing.Core.RandomVariables.Metadata{Dict{Turing.Core.RandomVariables.VarName{:phi},Int64},Array{Dirichlet{Float64},1},Array{Turing.Core.RandomVariables.VarName{:phi},1},Array{Float64,1},Array{Set{Turing.Selector},1}}}},Float64},Turing.Sampler{HMC{Turing.Core.TrackerAD,(),AdvancedHMC.Adaptation.UnitEuclideanMetric}},Turing.Model{Tuple{:theta,:phi},Tuple{:K,:V,:M,:N,:w,:doc,:beta,:alpha,:T},getfield(Main, Symbol("###inner_function#376#8")),NamedTuple{(:K, :V, :M, :N, :w, :doc, :beta, :alpha, :T),Tuple{Int64,Int64,Int64,Int64,Array{Int64,1},Array{Int64,1},Array{Float64,1},Array{Float64,1},Type{Float64}}},NamedTuple{(:K, :V, :M, :N, :w, :doc, :beta, :alpha),NTuple{8,Symbol}}}},getfield(Turing.Inference, Symbol("#∂logπ∂θ#30")){Turing.Core.RandomVariables.VarInfo{NamedTuple{(:theta, :phi),Tuple{Turing.Core.RandomVariables.Metadata{Dict{Turing.Core.RandomVariables.VarName{:theta},Int64},Array{Dirichlet{Float64},1},Array{Turing.Core.RandomVariables.VarName{:theta},1},Array{Float64,1},Array{Set{Turing.Selector},1}},Turing.Core.RandomVariables.Metadata{Dict{Turing.Core.RandomVariables.VarName{:phi},Int64},Array{Dirichlet{Float64},1},Array{Turing.Core.RandomVariables.VarName{:phi},1},Array{Float64,1},Array{Set{Turing.Selector},1}}}},Float64},Turing.Sampler{HMC{Turing.Core.TrackerAD,(),AdvancedHMC.Adaptation.UnitEuclideanMetric}},Turing.Model{Tuple{:theta,:phi},Tuple{:K,:V,:M,:N,:w,:doc,:beta,:alpha,:T},getfield(Main, Symbol("###inner_function#376#8")),NamedTuple{(:K, :V, :M, :N, :w, :doc, :beta, :alpha, :T),Tuple{Int64,Int64,Int64,Int64,Array{Int64,1},Array{Int64,1},Array{Float64,1},Array{Float64,1},Type{Float64}}},NamedTuple{(:K, :V, :M, :N, :w, :doc, :beta, :alpha),NTuple{8,Symbol}}}}}, ::Array{Float64,1}, ::Array{Float64,1}) at /Users/hg344/.julia/packages/AdvancedHMC/DdEnk/src/hamiltonian.jl:47
 [22] #sample#10(::Bool, ::Bool, ::Function, ::Random.MersenneTwister, ::AdvancedHMC.Hamiltonian{AdvancedHMC.Adaptation.UnitEuclideanMetric{Float64},getfield(Turing.Inference, Symbol("#logπ#31")){Turing.Core.RandomVariables.VarInfo{NamedTuple{(:theta, :phi),Tuple{Turing.Core.RandomVariables.Metadata{Dict{Turing.Core.RandomVariables.VarName{:theta},Int64},Array{Dirichlet{Float64},1},Array{Turing.Core.RandomVariables.VarName{:theta},1},Array{Float64,1},Array{Set{Turing.Selector},1}},Turing.Core.RandomVariables.Metadata{Dict{Turing.Core.RandomVariables.VarName{:phi},Int64},Array{Dirichlet{Float64},1},Array{Turing.Core.RandomVariables.VarName{:phi},1},Array{Float64,1},Array{Set{Turing.Selector},1}}}},Float64},Turing.Sampler{HMC{Turing.Core.TrackerAD,(),AdvancedHMC.Adaptation.UnitEuclideanMetric}},Turing.Model{Tuple{:theta,:phi},Tuple{:K,:V,:M,:N,:w,:doc,:beta,:alpha,:T},getfield(Main, Symbol("###inner_function#376#8")),NamedTuple{(:K, :V, :M, :N, :w, :doc, :beta, :alpha, :T),Tuple{Int64,Int64,Int64,Int64,Array{Int64,1},Array{Int64,1},Array{Float64,1},Array{Float64,1},Type{Float64}}},NamedTuple{(:K, :V, :M, :N, :w, :doc, :beta, :alpha),NTuple{8,Symbol}}}},getfield(Turing.Inference, Symbol("#∂logπ∂θ#30")){Turing.Core.RandomVariables.VarInfo{NamedTuple{(:theta, :phi),Tuple{Turing.Core.RandomVariables.Metadata{Dict{Turing.Core.RandomVariables.VarName{:theta},Int64},Array{Dirichlet{Float64},1},Array{Turing.Core.RandomVariables.VarName{:theta},1},Array{Float64,1},Array{Set{Turing.Selector},1}},Turing.Core.RandomVariables.Metadata{Dict{Turing.Core.RandomVariables.VarName{:phi},Int64},Array{Dirichlet{Float64},1},Array{Turing.Core.RandomVariables.VarName{:phi},1},Array{Float64,1},Array{Set{Turing.Selector},1}}}},Float64},Turing.Sampler{HMC{Turing.Core.TrackerAD,(),AdvancedHMC.Adaptation.UnitEuclideanMetric}},Turing.Model{Tuple{:theta,:phi},Tuple{:K,:V,:M,:N,:w,:doc,:beta,:alpha,:T},getfield(Main, Symbol("###inner_function#376#8")),NamedTuple{(:K, :V, :M, :N, :w, :doc, :beta, :alpha, :T),Tuple{Int64,Int64,Int64,Int64,Array{Int64,1},Array{Int64,1},Array{Float64,1},Array{Float64,1},Type{Float64}}},NamedTuple{(:K, :V, :M, :N, :w, :doc, :beta, :alpha),NTuple{8,Symbol}}}}}, ::AdvancedHMC.StaticTrajectory{AdvancedHMC.Leapfrog{Float64}}, ::Array{Float64,1}, ::Int64, ::AdvancedHMC.Adaptation.NoAdaptation, ::Int64) at /Users/hg344/.julia/packages/AdvancedHMC/DdEnk/src/sampler.jl:34
 [23] #sample at ./int.jl:0 [inlined]
 [24] #sample at ./none:0 [inlined] (repeats 2 times)
 [25] #steps!#28(::Random.MersenneTwister, ::Bool, ::Function, ::Turing.Model{Tuple{:theta,:phi},Tuple{:K,:V,:M,:N,:w,:doc,:beta,:alpha,:T},getfield(Main, Symbol("###inner_function#376#8")),NamedTuple{(:K, :V, :M, :N, :w, :doc, :beta, :alpha, :T),Tuple{Int64,Int64,Int64,Int64,Array{Int64,1},Array{Int64,1},Array{Float64,1},Array{Float64,1},Type{Float64}}},NamedTuple{(:K, :V, :M, :N, :w, :doc, :beta, :alpha),NTuple{8,Symbol}}}, ::Turing.Sampler{HMC{Turing.Core.TrackerAD,(),AdvancedHMC.Adaptation.UnitEuclideanMetric}}, ::Turing.Core.RandomVariables.VarInfo{NamedTuple{(:theta, :phi),Tuple{Turing.Core.RandomVariables.Metadata{Dict{Turing.Core.RandomVariables.VarName{:theta},Int64},Array{Dirichlet{Float64},1},Array{Turing.Core.RandomVariables.VarName{:theta},1},Array{Float64,1},Array{Set{Turing.Selector},1}},Turing.Core.RandomVariables.Metadata{Dict{Turing.Core.RandomVariables.VarName{:phi},Int64},Array{Dirichlet{Float64},1},Array{Turing.Core.RandomVariables.VarName{:phi},1},Array{Float64,1},Array{Set{Turing.Selector},1}}}},Float64}, ::Array{Turing.Utilities.Sample,1}) at /Users/hg344/.julia/dev/Turing/src/inference/hmc.jl:466
 [26] (::getfield(Turing.Inference, Symbol("#kw##steps!")))(::NamedTuple{(:rng, :verbose),Tuple{Random.MersenneTwister,Bool}}, ::typeof(Turing.Inference.steps!), ::Turing.Model{Tuple{:theta,:phi},Tuple{:K,:V,:M,:N,:w,:doc,:beta,:alpha,:T},getfield(Main, Symbol("###inner_function#376#8")),NamedTuple{(:K, :V, :M, :N, :w, :doc, :beta, :alpha, :T),Tuple{Int64,Int64,Int64,Int64,Array{Int64,1},Array{Int64,1},Array{Float64,1},Array{Float64,1},Type{Float64}}},NamedTuple{(:K, :V, :M, :N, :w, :doc, :beta, :alpha),NTuple{8,Symbol}}}, ::Turing.Sampler{HMC{Turing.Core.TrackerAD,(),AdvancedHMC.Adaptation.UnitEuclideanMetric}}, ::Turing.Core.RandomVariables.VarInfo{NamedTuple{(:theta, :phi),Tuple{Turing.Core.RandomVariables.Metadata{Dict{Turing.Core.RandomVariables.VarName{:theta},Int64},Array{Dirichlet{Float64},1},Array{Turing.Core.RandomVariables.VarName{:theta},1},Array{Float64,1},Array{Set{Turing.Selector},1}},Turing.Core.RandomVariables.Metadata{Dict{Turing.Core.RandomVariables.VarName{:phi},Int64},Array{Dirichlet{Float64},1},Array{Turing.Core.RandomVariables.VarName{:phi},1},Array{Float64,1},Array{Set{Turing.Selector},1}}}},Float64}, ::Array{Turing.Utilities.Sample,1}) at ./none:0
 [27] #sample#20(::Bool, ::Nothing, ::Int64, ::Nothing, ::Nothing, ::Random.MersenneTwister, ::Bool, ::Bool, ::Base.Iterators.Pairs{Union{},Union{},Tuple{},NamedTuple{(),Tuple{}}}, ::typeof(sample), ::Turing.Model{Tuple{:theta,:phi},Tuple{:K,:V,:M,:N,:w,:doc,:beta,:alpha,:T},getfield(Main, Symbol("###inner_function#376#8")),NamedTuple{(:K, :V, :M, :N, :w, :doc, :beta, :alpha, :T),Tuple{Int64,Int64,Int64,Int64,Array{Int64,1},Array{Int64,1},Array{Float64,1},Array{Float64,1},Type{Float64}}},NamedTuple{(:K, :V, :M, :N, :w, :doc, :beta, :alpha),NTuple{8,Symbol}}}, ::HMC{Turing.Core.TrackerAD,(),AdvancedHMC.Adaptation.UnitEuclideanMetric}) at /Users/hg344/.julia/dev/Turing/src/inference/hmc.jl:299
 [28] sample(::Turing.Model{Tuple{:theta,:phi},Tuple{:K,:V,:M,:N,:w,:doc,:beta,:alpha,:T},getfield(Main, Symbol("###inner_function#376#8")),NamedTuple{(:K, :V, :M, :N, :w, :doc, :beta, :alpha, :T),Tuple{Int64,Int64,Int64,Int64,Array{Int64,1},Array{Int64,1},Array{Float64,1},Array{Float64,1},Type{Float64}}},NamedTuple{(:K, :V, :M, :N, :w, :doc, :beta, :alpha),NTuple{8,Symbol}}}, ::HMC{Turing.Core.TrackerAD,(),AdvancedHMC.Adaptation.UnitEuclideanMetric}) at /Users/hg344/.julia/dev/Turing/src/inference/hmc.jl:248
 [29] top-level scope at none:0

```

### PR DESCRIPTION
**Pull request by mohamed82008**
Fixes #850.

### CODE REVIEW

**Comment by mohamed82008** [comment]
I will add some tests for Dirichlet and PDMatDistribution Tracker AD.

File: src/core/ad.jl [hunk 1]
```diff
@@ -433,3 +429,38 @@ MvNormal(m::TrackedVector{<:Real}, A::UniformScaling{<:Real}) = MvNormal(m, A.λ
 
 # zero mean,, constant variance
 MvNormal(d::Int, σ::TrackedReal{<:Real}) = MvNormal(zeros(d), σ)
+
+import Bijectors: link, invlink
+using Bijectors: Dirichlet, PDMatDistribution
+
+for F in (:link, :invlink)
+    @eval begin
+        $F(dist::Dirichlet, x::Tracker.TrackedArray, ::Type{Val{proj}}=Val{true}) where {proj} = Tracker.track($F, dist, x, Val{proj})
```
Comment (willtebbutt) [inline src/core/ad.jl]: Line too long?

File: src/core/ad.jl [hunk 2]
```diff
@@ -433,3 +429,38 @@ MvNormal(m::TrackedVector{<:Real}, A::UniformScaling{<:Real}) = MvNormal(m, A.λ
...
+        Tracker.@grad function $F(dist::Dirichlet, x::Tracker.TrackedArray, ::Type{Val{proj}}) where {proj}
+            x_data = Tracker.data(x)
+            y = $F(dist, x_data, Val{proj})
+            return  y,
+                    Δ -> begin
+                        out = ForwardDiff.jacobian(x -> $F(dist, x, Val{proj}), x_data)' * Δ
```
Comment (willtebbutt) [inline src/core/ad.jl]: Please sort out code formatting here. The following would be more inline with the style guide IMHO:
```julia
return y, Δ -> begin
    out = ForwardDiff.jacobian(x -> $F(dist, x, Val{proj}), x_data)' * Δ
    return (nothing, out, nothing)
end
```

File: src/core/ad.jl [hunk 3]
```diff
@@ -433,3 +429,38 @@ MvNormal(m::TrackedVector{<:Real}, A::UniformScaling{<:Real}) = MvNormal(m, A.λ
...
+                        return (nothing, out, nothing)
+                    end
+        end
+    end
+end
+
+Base.:*(x::Adjoint{T, <:AbstractMatrix{T}}, y::TrackedVector) where {T} = Tracker.track(*, x, y)
```
Comment (willtebbutt) [inline src/core/ad.jl]: Consider re-writing as
```julia
Base.:*(x::Adjoint{T, <:AbstractMatrix{T}} where {T}, y::TrackedVector) = Tracker.track(*, x, y)
```
for cleanliness.

**Comment by mohamed82008** [comment]
There are some failed tests, supposedly not related to this PR. After fixing them, this can be merged.

**Comment by mohamed82008** [comment]
It seems the same test failures are on master. It's probably better to fix them in a separate PR.

**Comment by mohamed82008** [comment]
I made a separate PR for fixing the remaining tests. I think we can go ahead and merge this PR if there are no more comments.

**Pull request merged**